src/basic/chattr-util.c

```diff
@@ -172,3 +173,61 @@ int read_attr_at(int dir_fd, const char *path, unsigned *ret) {
 
         return read_attr_fd(fd, ret);
 }
+
+int get_proj_id(int fd, uint32_t *ret) {
+        struct fsxattr attrs;
+
+        assert(ret);
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        *ret = attrs.fsx_projid;
+
+        return 0;
+}
+
+int set_proj_id(int inode_fd, uint32_t proj_id) {
+        struct fsxattr attrs;
+        _cleanup_close_ int fd = -EBADF;
+
+        fd = fd_reopen(inode_fd, O_RDONLY|O_CLOEXEC);
+        if (fd < 0)
+                return fd;
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        if (attrs.fsx_projid == proj_id)
+                return 0;
+
+        attrs.fsx_projid = proj_id;
+
+        return RET_NERRNO(ioctl(fd, FS_IOC_FSSETXATTR, &attrs));
+}
+
+static int set_proj_id_cb(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+
```

Comment:
this callback gets called a bunch of times for each inode possibly.

please condition on RECURSE_DIR_ENTER and RECURSE_DIR_ENTRY, so that we only adjust this once per inode.